vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+    for (auto edge : node_out_edges(src_node)) {
+        if (edge_sink_node(edge) == sink_node) {
+            return edge;
+        }
+    }
+
+    //Not found
+    return RREdgeId::INVALID();
+}
+
+RRNodeId RRGraph::find_node(short x, short y, t_rr_type type, int ptc, e_side side) const {
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
+    }
+    size_t itype = type;
+    size_t iside = side;
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((x < 0)                                     /* See if x is smaller than the index of first element */
+        || (size_t(x) > node_lookup_.size() - 1)) { /* See if x is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((y < 0)                                        /* See if y is smaller than the index of first element */
+        || (size_t(y) > node_lookup_[x].size() - 1)) { /* See if y is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    /* itype is always larger than -1, we can skip checking */
+    if (itype > node_lookup_[x][y].size() - 1) { /* See if type is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((ptc < 0)                                                 /* See if ptc is smaller than the index of first element */
+        || (size_t(ptc) > node_lookup_[x][y][type].size() - 1)) { /* See if ptc is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    /* Check if x, y, type and ptc, side is valid */
+    /* iside is always larger than -1, we can skip checking */
+    if (iside > node_lookup_[x][y][type][ptc].size() - 1) { /* See if side is large than the index of last element */
+        /* Return a zero range! */
+        return RRGRAPH_OPEN_NODE_ID;
+    }
+
+    return node_lookup_[x][y][itype][ptc][iside];
+}
+
+RRGraph::node_range RRGraph::find_nodes(short x, short y, t_rr_type type, int ptc) const {
+    if (!valid_fast_node_lookup()) {
+        build_fast_node_lookup();
+    }
+
+    /* IMPORTANT!!! 
+     * check x, y, type, ptc if they are in the range
+     * if not, return a zero range 
+     */
+    size_t itype = type;
+
+    /* Check if x, y, type and ptc, side is valid */
+    if ((x < 0)                                     /* See if x is smaller than the index of first element */
+        || (size_t(x) > node_lookup_.size() - 1)) { /* See if x is large than the index of last element */
+        /* Return a zero range! */
+        const std::vector<RRNodeId> zero_range = std::vector<RRNodeId>();
```

Comment:
I don't think this is safe, since you're holding iterators to a locally variable which will go out of scope and be destructed before the range is accessed by the caller.

Instead we probably want to do something like:
```cpp
//Initialize iterators to a valid object, but zero sized range
auto begin = node_lookup_[0][0][SOURCE][0][0].begin();
auto end = node_lookup_[0][0][SOURCE][0][0].begin();

//Check all the bounds
//...

if (all bounds correct) {
    //Get the real bounds
    begin = node_lookup_[x][y][type][ptc][NUM_SIDES].begin()
    end = node_lookup_[x][y][type][ptc][NUM_SIDES].end()
}

return vtr::make_range(begin, end);
```

Reply:
I removed this function. And we will see if we need in downstream codes.